scipy/sparse/linalg/isolve/iterative.py

```diff
@@ -220,65 +226,70 @@ def bicg(A, b, x0=None, tol=1e-5, maxiter=None, M=None, callback=None, atol=None
                '``scipy.sparse.linalg.LinearOperator``.')
 @non_reentrant()
 def bicgstab(A, b, x0=None, tol=1e-5, maxiter=None, M=None, callback=None, atol=None):
-    A, M, x, b, postprocess = make_system(A, M, x0, b)
-
-    n = len(b)
+    assert A.shape[0] == A.shape[1] and A.shape[0] == len(b), "The size of the matrix and the right-hand side does not match."
```

Comment:
```suggestion
    if not (A.shape[0] == A.shape[1] and A.shape[0] == len(b)):
    	raise ValueError("The size of the matrix and the right-hand side does not match.")
```
`assert` is generally regarded as not good practice because using `python -OO` strips them away, see : https://mail.python.org/pipermail/scipy-dev/2021-March/024590.html

I would suggest changing them in the other lines as well.

Reply:
Thank you a lot for sharing your suggestions and experiences in Python, it helps me a lot. I'll try to modify it.